#650 Make HttpClient support compressed responses

kewne opened the pull request.

rhart commented on ratpack-core/src/test/groovy/ratpack/http/client/HttpProxySpec.groovy:
```diff
@@ -193,4 +196,32 @@ transfer-encoding: chunked
+    otherApp {
+      get("foo") {
+        response.headers.add(HttpHeaderConstants.CONTENT_ENCODING, "gzip")
+        response.send(stringBuf.toByteArray())
```
I don't think we need to do this.  If we add the accept-encoding header to the request then the "otherApp" should deal with the zipping and adding the correct header.

rhart commented:
Thanks!  Do you fancy looking at the streaming too? :)

kewne commented:
Sure, not sure why I missed it...

kewne commented:
Hmmm, I also just realized this breaks ```receivedResponse.send(response)```, so it'll need some more work... Good news is streaming is working too :)

alkemist commented:
Looks like there are some failing tests in ratpack-core with this change.

alkemist commented:
@kewne is this “done” if so, would you mind looking at the failing tests?

kewne commented:
Sure, I just haven't had the time to go through this yet.
From what I've seen, the ```HttpContentDecompressor``` I've added also changes the ```Content-Encoding``` header. Since ```DefaultHttpClient``` is used in the tests too, the tests fails because the headers don't match...
I'll try to think of a way to get around this, let me know if you have any ideas.

kewne commented:
Finally got some time to finish this...

rhart commented:
Great!  Shout if you need anything

kewne commented:
I think it's finished.
In the end, I decided to add a ```RequestSpec.decompressResponse``` method to enable/disable on a per request basis, like you suggested.

Let me know if there's anything else I should add.

rhart commented:
Awesome.  I will take a look soon as poss.

rhart commented on ratpack-core/src/main/java/ratpack/http/client/internal/ContentAggregatingRequestAction.java:
```diff
@@ -45,6 +46,9 @@ public ContentAggregatingRequestAction(Action<? super RequestSpec> requestConfig
 
   @Override
   protected void addResponseHandlers(ChannelPipeline p, Fulfiller<? super ReceivedResponse> fulfiller) {
+    if (requestSpecBacking.isDecompressResponse()) {
```
Can this go in `RequestActionSupport`

rhart commented on ratpack-core/src/main/java/ratpack/http/client/internal/ContentStreamingRequestAction.java:
```diff
@@ -61,6 +62,9 @@ public ContentStreamingRequestAction(Action<? super RequestSpec> requestConfigur
 
   @Override
   protected void addResponseHandlers(ChannelPipeline p, Fulfiller<? super StreamedResponse> fulfiller) {
+    if (requestSpecBacking.isDecompressResponse()) {
```
Can this go in `RequestActionSupport`

kewne replied:
It can, great suggestion. I'll take care of it.

rhart commented on ratpack-core/src/main/java/ratpack/http/client/internal/RequestActionSupport.java:
```diff
@@ -54,7 +54,7 @@
   private final int port;
   private final MutableHeaders headers;
-  private final RequestSpecBacking requestSpecBacking;
+  final RequestSpecBacking requestSpecBacking;
```
We shouldn't need to do this if adding the decompressor is done in this class

rhart commented on ratpack-core/src/main/java/ratpack/http/client/internal/RequestSpecBacking.java:
```diff
@@ -51,6 +52,7 @@ public RequestSpecBacking(MutableHeaders headers, URI uri, ByteBufAllocator byte
     this.byteBufAllocator = byteBufAllocator;
     this.requestParams = requestParams;
     this.bodyByteBuf = byteBufAllocator.buffer(0, 0);
+    this.decompressResponse = false;
```
I think we should default to true, this will be the most common case

kewne replied:
That breaks backwards compatibility and a few tests, so I wanted to avoid it :stuck_out_tongue:
If it's not a big concern to you (understandably, it's not even 1.0 at this point) let me know.

rhart replied:
I would say now is the time, before 1.0, to get this how we want.  Let's take advantage of not being concerned about backwards compatibility while we still can :)

alkemist replied:
+1



> On 27 May 2015, at 21:15, Rus <notifications@github.com> wrote:
> 
> In ratpack-core/src/main/java/ratpack/http/client/internal/RequestSpecBacking.java:
> 
> > @@ -51,6 +52,7 @@ public RequestSpecBacking(MutableHeaders headers, URI uri, ByteBufAllocator byte
> >      this.byteBufAllocator = byteBufAllocator;
> >      this.requestParams = requestParams;
> >      this.bodyByteBuf = byteBufAllocator.buffer(0, 0);
> > +    this.decompressResponse = false;
> I would say now is the time, before 1.0, to get this how we want. Let's take advantage of not being concerned about backwards compatibility while we still can :)
> 
> —
> Reply to this email directly or view it on GitHub.
>

rhart commented on ratpack-core/src/test/groovy/ratpack/http/client/HttpClientSmokeSpec.groovy:
```diff
@@ -517,4 +519,73 @@ BAR
+
+    otherApp {
+      get("foo") {
+        response.headers.add(CONTENT_ENCODING, "gzip")
```
We shouldn't need to do this or the GZipping.  If the client sends `Accept-Encoding` of `gzip` the response from this app should automatically get compressed

kewne replied:
I couldn't find the code where this is being done, can you give me some pointers?

One thing I'm worried about is that, by not making explicit that the response is compressed, ```otherApp```'s response could be uncompressed, which would make the test pass incorrectly.

rhart replied:
This is where it's being done https://github.com/ratpack/ratpack/blob/877b7a2b6e465944601712a92410525314c3a5da/ratpack-core/src/main/java/ratpack/server/internal/SmartHttpContentCompressor.java

I can't seem to find any tests specifically for this but there are some for file compression.

I do take your point though.  My concern is that I don't know what happens if the client sends the `Accept-Encoding` header in this case, will otherApp try to apply compression to an already compressed response??  I guess not or your tests wouldn't be passing.  Maybe we need to put other assertions in along the way to make sure the behaviour is as we expect.

kewne replied:
Ok, I actually tried setting the header to do compression which, oddly enough, disables compression.
It worked when I just set ```Accept-Encoding``` on the client side and left the server side alone.

rhart commented on ratpack-core/src/test/groovy/ratpack/http/client/HttpClientSmokeSpec.groovy:
```diff
@@ -517,4 +519,73 @@ BAR
+    handlers {
+      get { HttpClient httpClient ->
+        httpClient.request(otherAppUrl("foo")) { RequestSpec rs ->
+          rs.headers.set("Accept-Encoding", "compress, gzip")
```
If decompress is on, I wonder whether we should automatically set these headers if they're not already set

rhart commented:
Sorry for taking so long to review.  If you're happy to address the comments that'd be great, otherwise I'll take a look as part of merging.  Thanks again for this.

kewne commented:
Not sure if you guys receive notifications when I push so just making sure.
I think I've addressed all of your comments but let me know if there's some other change to be done.

rhart commented:
Great!  Can you add yourself to the contributors list (https://github.com/ratpack/ratpack/blob/master/ratpack-manual/src/content/chapters/99-about-the-project.md) and I'll take a look over the weekend

rhart commented:
Thanks again for this @kewne, great stuff!
